### PR TITLE
storage,roachpb: allow reading MVCCValueHeader data via SQL

### PR DESCRIPTION
**Pull request by stevendanna**
Logical Data Replication (LDR) and Import now both write values into
the MVCCValueHeader of keys. We would like to be able to
read these values from SQL via new system columns. In the case of LDR this will allow the
SQL-write-path to continue to function even after the OriginTimetamp
column is moved out of the user's table schema and into the
MVCCValueHeader.

We make that possible by:

- Adding a ReturnRawMVCCValue option to GetRequest, ScanRequest, and
ReverseScanRequest that returns values with the extended encoding rather than
stripping off the MVCCValueHeader.

- Modifying the functions on roachpb.Value to understand the extended
  encoding may be in place and return the correct version.

- Modifying the column and rowfetchers to make requests with this new option
  when the newly added system columns are added.

Release note: None
Epic: none

### CODE REVIEW

**Comment by stevendanna** [comment]
@nvanbenschoten @dt I have a gazillion data driven SQL tests that probably need to be rewritten and some new tests to push, but if y'all have opinions about this general strategy before I go and do the test rewriting let me know.

**Comment by stevendanna** [comment]
I've marked this ready for review even though I haven't updated the SQL level tests.  The first commit can probably be reviewed and merged separately here: https://github.com/cockroachdb/cockroach/pull/131348

**Comment by stevendanna** [comment]
The test rewrites are going to make this huge.  The most important things to review are:

```
 pkg/ccl/changefeedccl/cdcevent/event.go            |    5 +-
 pkg/sql/catalog/catalog.go                         |    2 +-
 pkg/sql/catalog/catpb/enum.proto                   |    6 +
 pkg/sql/catalog/colinfo/system_columns.go          |   44 +
 pkg/sql/colfetcher/cfetcher.go                     |   67 +-
 pkg/sql/colfetcher/cfetcher_setup.go               |   15 +
 pkg/sql/colfetcher/colbatch_direct_scan.go         |    1 +
 pkg/sql/colfetcher/colbatch_scan.go                |    1 +
 pkg/sql/colfetcher/index_join.go                   |    1 +
 pkg/sql/crdb_internal_test.go                      |   38 +
 pkg/sql/distsql_plan_changefeed.go                 |    5 +-
 pkg/sql/row/fetcher.go                             |   52 +-
 pkg/sql/row/kv_batch_fetcher.go                    |   12 +-
 pkg/sql/row/kv_batch_streamer.go                   |    2 +-
 pkg/sql/row/kv_fetcher.go                          |    8 +-
```

File: pkg/roachpb/data.go [hunk 1]
```diff
@@ -346,7 +354,20 @@ func (v *Value) ShallowClone() *Value {
 
 // IsPresent returns true if the value is present (existent and not a tombstone).
 func (v *Value) IsPresent() bool {
-	return v != nil && len(v.RawBytes) != 0
+	if v == nil || len(v.RawBytes) == 0 {
+		return false
+	}
+	// TODO(ssd): This is a bit awkward because this is the right thing to
```
Comment (tbg) [inline pkg/roachpb/data.go:360]: Can you address this TODO before merging?

**Review by tbg — COMMENTED** [review]
Only reviewed the `roachpb` commit and it looks as expected, though I'd appreciate a thorough look line-by-line and at the test by @miraradeva (thank you!)
There are two TODOs that seem like they should be addressed (in their own commits).

**Comment by stevendanna** [comment]
I opened a PR with just the roachpb.Value items here: https://github.com/cockroachdb/cockroach/pull/132088

**Review by michae2 — CHANGES_REQUESTED** [review]
I looked at the last two commits (sans testcases). Only noticed a few small things.

> Reviewed 1 of 19 files at r7, 15 of 125 files at r10, 3 of 3 files at r11, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/130957#-:-O8k2GFa-3uSt_ThuBEx:bgssu4c)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @dt, @miraradeva, and @stevendanna)

___
*[`pkg/sql/colfetcher/cfetcher.go` line 864 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/130957#-O8k0Nz6BP8OktyKWAZO:-O8k0Nz6BP8OktyKWAZP:b-1kwu2s) ([raw file](https://github.com/cockroachdb/cockroach/blob/b664078e9f382af787064116501c0c38dd6a847e/pkg/sql/colfetcher/cfetcher.go#L864)):*
> ```Go
> 
> 			if vh, err := cf.machine.nextKV.Value.GetMVCCValueHeader(); err == nil {
> 				cf.table.rowLastOriginID = int(vh.OriginID)
> ```

Similar to another comment, I think we should only be setting rowLastOriginID if rowLastOriginTimestamp is less than vh.OriginTimestamp.
___
*[`pkg/sql/colfetcher/cfetcher.go` line 939 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/130957#-O8k0hFO8PZCrE0jdttI:-O8k0hFO8PZCrE0jdttJ:b-9dhwii) ([raw file](https://github.com/cockroachdb/cockroach/blob/b664078e9f382af787064116501c0c38dd6a847e/pkg/sql/colfetcher/cfetcher.go#L939)):*
> ```Go
> 
> 			if vh, err := kv.Value.GetMVCCValueHeader(); err == nil {
> 				cf.table.rowLastOriginID = int(vh.OriginID)
> ```

same comment here
___
*[`pkg/sql/row/fetcher.go` line 910 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/130957#-O8jrPiCE_TpIn0ZzH0I:-O8jrPiCE_TpIn0ZzH0J:b-8ftljl) ([raw file](https://github.com/cockroachdb/cockroach/blob/b664078e9f382af787064116501c0c38dd6a847e/pkg/sql/row/fetcher.go#L910)):*
> ```Go
> 	}
> 	if vh, err := kv.Value.GetMVCCValueHeader(); err == nil {
> 		if table.rowLastOriginID < int(vh.OriginID) {
> ```

Is this right? If there are multiple column families, and the most-recently-updated one has a higher origin ID than the others, seems like we would still want to keep the origin ID of the most-recently-updated one.

In other words, shouldn't _both_ origin ID and origin timestamp be decided by the column family with the latest origin timestamp?
___
*[`pkg/sql/row/fetcher.go` line 1334 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/130957#-O8k221YBWbDbQ-VVq1P:-O8k221YBWbDbQ-VVq1Q:b2609p4) ([raw file](https://github.com/cockroachdb/cockroach/blob/d4b1e48271e3c76bd7fa76863078cab3c382f3ef/pkg/sql/row/fetcher.go#L1334)):*
> ```Go
> 			table.row[table.originTimestampOutputIdx] = rowenc.EncDatum{Datum: dec}
> 		} else {
> 			table.row[table.originTimestampOutputIdx] = rowenc.EncDatum{Datum: tree.DNull}
> ```

can also use `rowenc.NullEncDatum()`
___
*[`pkg/sql/row/kv_batch_streamer.go` line 109 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/130957#-O8jubuJ23qxwkfXFqAo:-O8jubuJ23qxwkfXFqAp:b21hsuu) ([raw file](https://github.com/cockroachdb/cockroach/blob/b664078e9f382af787064116501c0c38dd6a847e/pkg/sql/row/kv_batch_streamer.go#L109)):*
> ```Go
> 	}
> 	// TODO(yuzefovich): consider supporting COL_BATCH_RESPONSE scan format.
> 	reqs := spansToRequests(spans, kvpb.BATCH_RESPONSE, false /* reverse */, false /* rawMVCCValues */, f.lockStrength, f.lockDurability, reqsScratch)
> ```

rawMVCCValues needs to be plumbed into the streamer, too, in order for lookup joins to be able to return these columns.

**Review by stevendanna — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/130957)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @dt, @michae2, @miraradeva, and @tbg)

___
*[`pkg/roachpb/data.go` line 279 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/130957#gh-1789832135:-O8kbtEP7MNBFxeugd_2:bbocgdq) ([raw file](https://github.com/cockroachdb/cockroach/blob/0d376ee117e1d1b1958c4f1c0175c47f02e89a62/pkg/roachpb/data.go#L279)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

in the commit message you say

> - checksum(): Currently returns 0 if it finds an extended
  encoding. This should be improved in a follow-up commit.

What improvement were you thinking of? Can you add the follow-up commit?
</blockquote></details>

I've addressed this in a follow-up commit that can be seen over on the other PR: https://github.com/cockroachdb/cockroach/pull/132088. Apologies for the 2 PRs, keeping the tests passing on this branch is a bit much.
___
*[`pkg/roachpb/data.go` line 360 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/130957#gh-1789827579:-O8kcHNn-_9q6oIHfqXs:b-1eczy9) ([raw file](https://github.com/cockroachdb/cockroach/blob/0d376ee117e1d1b1958c4f1c0175c47f02e89a62/pkg/roachpb/data.go#L360)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

Can you address this TODO before merging?
</blockquote></details>

How much would you hate it if I addressed this as part of https://github.com/cockroachdb/cockroach/issues/131664?
___
*[`pkg/sql/colfetcher/cfetcher.go` line 864 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/130957#-O8k0Nz6BP8OktyKWAZO:-O8kc_FqBSENsIqEFatJ:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b664078e9f382af787064116501c0c38dd6a847e/pkg/sql/colfetcher/cfetcher.go#L864)):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

Similar to another comment, I think we should only be setting rowLastOriginID if rowLastOriginTimestamp is less than vh.OriginTimestamp.
</blockquote></details>

Done.
___
*[`pkg/sql/colfetcher/cfetcher.go` line 939 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/130957#-O8k0hFO8PZCrE0jdttI:-O8kcZPX7DY_IcAIOP1k:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b664078e9f382af787064116501c0c38dd6a847e/pkg/sql/colfetcher/cfetcher.go#L939)):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

same comment here
</blockquote></details>

Done.
___
*[`pkg/sql/row/fetcher.go` line 910 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/130957#-O8jrPiCE_TpIn0ZzH0I:-O8kc_rlBPzIKthFPzmZ:b-u0c0i) ([raw file](https://github.com/cockroachdb/cockroach/blob/b664078e9f382af787064116501c0c38dd6a847e/pkg/sql/row/fetcher.go#L910)):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

Is this right? If there are multiple column families, and the most-recently-updated one has a higher origin ID than the others, seems like we would still want to keep the origin ID of the most-recently-updated one.

In other words, shouldn't _both_ origin ID and origin timestamp be decided by the column family with the latest origin timestamp?
</blockquote></details>

The semantics that this should have for multiple column families almost surely needs more thought. Currently, no one writing this value header can use multiple column families.  However, I agree that it is better to have this closer to correct. Towards that end I updated the semantics of both columns:

- origin_timestamp is now the largest origin_timestamp but with the caveat that it returns NULL if there is a column family with no OriginTimestamp and a MVCC timestamp larger than the largest origin_timestamp.

- origin_id is now the largest OriginID of the row with the largest OriginTimestamp set. I didn't make this conditional on whether there is a row with a larger MVCC timestamp.  We could, but right now this column is really just for debugging and it isn't clear which semantics are most useful for us yet.
___
*[`pkg/sql/row/fetcher.go` line 1334 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/130957#-O8k221YBWbDbQ-VVq1P:-O8kf67F9OQau1Szf6IR:bkwaqa7) ([raw file](https://github.com/cockroachdb/cockroach/blob/d4b1e48271e3c76bd7fa76863078cab3c382f3ef/pkg/sql/row/fetcher.go#L1334)):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

can also use `rowenc.NullEncDatum()`
</blockquote></details>

Oh nice.
___
*[`pkg/sql/row/kv_batch_streamer.go` line 109 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/130957#-O8jubuJ23qxwkfXFqAo:-O8kf8Ht6SpoaeKd8Zd3:byumxhe) ([raw file](https://github.com/cockroachdb/cockroach/blob/b664078e9f382af787064116501c0c38dd6a847e/pkg/sql/row/kv_batch_streamer.go#L109)):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

rawMVCCValues needs to be plumbed into the streamer, too, in order for lookup joins to be able to return these columns.
</blockquote></details>

Done. And I added a test that hits the index join case.

**Review by tbg — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/130957)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @dt, @michae2, @miraradeva, and @stevendanna)

___
*[`pkg/roachpb/data.go` line 360 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/130957#gh-1789827579:-O8l9umI91vtaHgkPer6:bloozyz) ([raw file](https://github.com/cockroachdb/cockroach/blob/0d376ee117e1d1b1958c4f1c0175c47f02e89a62/pkg/roachpb/data.go#L360)):*
<details><summary><i>Previously, stevendanna (Steven Danna) wrote…</i></summary><blockquote>

How much would you hate it if I addressed this as part of https://github.com/cockroachdb/cockroach/issues/131664?
</blockquote></details>

Sounds good, if you seem convinced it's not going to cause issues until then?

**Comment by stevendanna** [comment]
> Sounds good, if you seem convinced it's not going to cause issues until then?

So far the only time I've hit this issue is when I turn on "always send back raw MVCC values" to callers who didn't opt into them. Now, perhaps there is test area that I haven't randomly hit in any of my pushes or local test runs, but I think right now the two callers who send requests with this option set will correctly handle the responses.

**Comment by miraradeva** [comment]
> The semantics that this should have for multiple column families almost surely needs more thought.

I understand that multiple column families are not a typical use case for LDR, but are they another case where the SQL-only vs KV-pushdown behaviors could result in different outcomes? Essentially because, with KV pushdown, each of the key-value pairs corresponding to different column families of a single row could have different origin id/timestamp, while with SQL, there is a single origin id/timestamp for the whole row. E.g. if the max origin timestamp of the row is 100 but one column family only had a max origin timestamp of 50, would we ignore an update for this column family with origin timestamp of 80?

**Review by michae2 — APPROVED** [review]
:lgtm: Nice work!

> Reviewed 14 of 125 files at r13, 3 of 3 files at r14, 9 of 9 files at r16, 1 of 1 files at r17, 1 of 1 files at r18, 3 of 3 files at r19, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/130957#-:-O8mJcGL9BzQ19nKVBDD:bcmw7ss)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @dt, @miraradeva, @stevendanna, and @tbg)

**Comment by stevendanna** [comment]
bors r=michae2

**Pull request merged**